### ISSUE TITLE
How to repeatedly save the best validation loss?

### ISSUE THREAD

**Issue by wbrickner** [Tracel-AI/burn#738]
Is there a way to do this in burn?

I want to save (to a file) the model / parameters from the step with the minimum validation loss.

**Comment by nathanielsimard** [comment]
For now, the only way to do this is to save all models and keep the one with the best validation loss for the detected epoch. There is significant work to be done in burn-train to support this gracefully, namely making collected metrics available, like a database, so that logic could be built on top.

**Comment by wbrickner** [comment]
How would this be done now? Can you provide some sort of example?

**Comment by nathanielsimard** [comment]
```rust
    let learner = LearnerBuilder::new(artifact_dir)
        .with_file_checkpointer(usize::MAX, CompactRecorder::new())
        .build(model, optim, lr_scheduler);
```

This will never delete checkpoints, so you will be able to chose the one with the best validation loss.
If you want to delete the checkpoints during training, you could create a cron job that reads the metrics saved on the disk and only keep the checkpoints with the best validation loss. I don't have any specific recommendations on implementing this for now :/

**Comment by antimora** [comment]
@nathanielsimard You worked on enhancements to support this. Should be close it now?

**Comment by nathanielsimard** [comment]
Yes, you can add an early stopping strategy:

```rust
.early_stopping(MetricEarlyStoppingStrategy::new::<LossMetric<B>>(
            Aggregate::Mean,
            Direction::Lowest,
            Split::Valid,
            StoppingCondition::NoImprovementSince { n_epochs: 1 },
        ))
```

and a checkpointing strategy:

```rust
ComposedCheckpointingStrategy::builder()
                    .add(KeepLastNCheckpoints::new(2))
                    .add(MetricCheckpointingStrategy::new::<LossMetric<B>>(
                        Aggregate::Mean,
                        Direction::Lowest,
                        Split::Valid,
                    ))
                    .build(),
```

So it's completed.
